File: app/src/main/java/app/revanced/bilibili/settings/fragments/BaseDynamicItemsFragment.java
package app.revanced.bilibili.settings.fragments;

import android.app.AlertDialog;
import android.os.Bundle;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import app.revanced.bilibili.meta.BottomItem;
import app.revanced.bilibili.settings.Settings;
import app.revanced.bilibili.utils.Constants;
import app.revanced.bilibili.utils.Utils;
import app.revanced.bilibili.widget.CheckBoxGroupPreference;

public abstract class BaseDynamicItemsFragment extends BiliRoamingBaseSettingFragment {
    private final Settings setting;
    private final List<BottomItem> allItems;
    private boolean hintShown = false;

    public BaseDynamicItemsFragment(String prefsXmlName, Settings setting, List<BottomItem> allItems) {
        super(prefsXmlName);
        this.setting = setting;
        this.allItems = allItems;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        CheckBoxGroupPreference preference = findPreference(setting.key);
        if (preference == null) return;
        Set<String> showingItems = setting.getStringSet();
        List<String> values = new ArrayList<>();
        List<String> entries = new ArrayList<>();
        for (BottomItem drawerItem : allItems) {
            values.add(drawerItem.id);
            entries.add(String.format("%s (%s)", drawerItem.name, drawerItem.id));
        }
        preference.setEntries(entries.toArray(new CharSequence[0]));
        preference.setEntryValues(values.toArray(new CharSequence[0]));
        Set<String> selectedValues = new HashSet<>();
        if (showingItems.size() == 1 && showingItems.contains(Constants.ALL_VALUE)) {
            selectedValues.addAll(values);
        } else {
            selectedValues.addAll(showingItems);
        }
        preference.setValues(selectedValues);
        preference.notifyChanged();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (hintShown) return;
        if (allItems.isEmpty()) {
            hintShown = true;
            new AlertDialog.Builder(requireContext())
                    .setMessage(Utils.getString("biliroaming_hint_open_from_mine"))
                    .setPositiveButton(Utils.getString("biliroaming_get_it"), null)
                    .show();
        }
    }
}

File: integrations/app/src/main/java/app/revanced/bilibili/settings/fragments/BaseDynamicItemsFragment.kt
package app.revanced.bilibili.settings.fragments

import android.app.AlertDialog
import android.os.Bundle
import app.revanced.bilibili.meta.BottomItem
import app.revanced.bilibili.settings.StringSetSetting
import app.revanced.bilibili.utils.Constants
import app.revanced.bilibili.utils.Utils
import app.revanced.bilibili.utils.constraintSize
import app.revanced.bilibili.widget.CheckBoxGroupPreference

abstract class BaseDynamicItemsFragment(
    private val setting: StringSetSetting,
    private val allItems: List<BottomItem>
) : BiliRoamingBaseSettingFragment() {
    private var hintShown = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val preference = findPreference<CheckBoxGroupPreference>(setting.key) ?: return
        val showingItems = setting()
        val values = mutableListOf<String>()
        val entries = mutableListOf<String>()
        for (item in allItems) {
            values.add(item.id)
            entries.add("${item.name} (${item.id})")
        }
        preference.entries = entries.toTypedArray<CharSequence>()
        preference.entryValues = values.toTypedArray<CharSequence>()
        val selectedValues = mutableSetOf<String>()
        if (showingItems.size == 1 && showingItems.contains(Constants.ALL_VALUE)) {
            selectedValues.addAll(values)
        } else {
            selectedValues.addAll(showingItems)
        }
        preference.values = selectedValues
        preference.notifyChanged()
    }

    override fun onResume() {
        super.onResume()
        if (hintShown) return
        if (allItems.isNotEmpty()) return
        hintShown = true
        AlertDialog.Builder(context)
            .setMessage(Utils.getString("biliroaming_hint_open_from_mine"))
            .setPositiveButton(Utils.getString("biliroaming_get_it"), null)
            .create().constraintSize().show()
    }
}
